ome_zarr/io.py

```diff
@@ -0,0 +1,150 @@
+"""Reading logic for ome-zarr.
+
+Primary entry point is the :func:`~ome_zarr.io.parse_url` method.
+"""
+
+import json
+import logging
+import os
+import posixpath
+from abc import ABC, abstractmethod
+from typing import Optional
+from urllib.parse import urlparse
+
+import dask.array as da
+import requests
+
+from .types import JSONDict
+
+LOGGER = logging.getLogger("ome_zarr.io")
+
+
+class BaseZarrLocation(ABC):
+    """
+    Base IO primitive for reading Zarr data.
+
+    No assumptions about the existence of the given path string are made.
+    Attempts are made to load various metadata files and cache them internally.
+    """
+
+    def __init__(self, path: str) -> None:
+        self.zarr_path: str = path.endswith("/") and path or f"{path}/"
+        self.zarray: JSONDict = self.get_json(".zarray")
+        self.zgroup: JSONDict = self.get_json(".zgroup")
+        self.__metadata: JSONDict = {}
+        self.__exists: bool = True
+        if self.zgroup:
+            self.__metadata = self.get_json(".zattrs")
+        elif self.zarray:
+            self.__metadata = self.get_json(".zattrs")
+        else:
+            self.__exists = False
+
+    def __repr__(self) -> str:
+        """Print the path as well as whether this is a group or an array."""
+        suffix = ""
+        if self.zgroup:
+            suffix += " [zgroup]"
+        if self.zarray:
+            suffix += " [zarray]"
+        return f"{self.zarr_path}{suffix}"
+
+    def exists(self) -> bool:
+        """Return true if zgroup or zarray metadata exists."""
+        return self.__exists
+
+    def is_zarr(self) -> Optional[JSONDict]:
+        """Return true if either zarray or zgroup metadata exists."""
```

Comment:
I don't quite get what's being communicated by "either" here but not for the previous.

Reply:
A node in the Zarr hierarchy must have one or the other.

Reply:
Aha, but that's not true for `exists`, it could be both? Okay.

Reply:
Pushed a removal of `is_zarr()` in favor of `exists()`